src/coprocessor/codec/statistics/fmsketch.rs

```diff
@@ -0,0 +1,186 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// FIXME: remove following later
+#![allow(dead_code)]
+
+use std::collections::HashSet;
+use std::hash::Hasher;
+
+use fnv::FnvHasher;
+use tipb::analyze;
+use coprocessor::codec::datum;
+use coprocessor::codec::datum::Datum;
+use coprocessor::codec::Result;
+
+pub struct FMSketch {
+    mask: u64,
+    max_size: usize,
+    hash_set: HashSet<u64>,
+}
+
+impl FMSketch {
+    pub fn new(max_size: usize) -> FMSketch {
+        FMSketch {
+            mask: 0,
+            max_size: max_size,
+            hash_set: HashSet::new(),
+        }
+    }
+
+    pub fn build(values: Vec<Datum>, max_size: usize) -> Result<FMSketch> {
+        let mut s = FMSketch::new(max_size);
+        for value in values {
+            try!(s.insert(value));
+        }
+        Ok(s)
+    }
+
+    fn merge(&mut self, rs: &FMSketch) {
+        if self.mask < rs.mask {
+            self.update_mask(rs.mask);
+        }
+        for key in &rs.hash_set {
+            self.insert_hash_value(*key)
+        }
+    }
+
+    fn update_mask(&mut self, mask: u64) {
+        self.mask = mask;
+        let to_remove: Vec<u64> = self.hash_set
+            .iter()
+            .filter(|x| (*x & self.mask) != 0)
+            .cloned()
+            .collect();
+        for x in &to_remove {
+            self.hash_set.remove(x);
+        }
+    }
+
+    // returns the ndv of the sketch
+    pub fn ndv(&self) -> i64 {
+        (self.mask + 1) as i64 * (self.hash_set.len() as i64)
+    }
+
+    fn insert_hash_value(&mut self, hash_val: u64) {
+        if (hash_val & self.mask) != 0 {
+            return;
+        }
+        self.hash_set.insert(hash_val);
+        if self.hash_set.len() > self.max_size {
+            let mask = (self.mask << 1) | 1;
+            self.update_mask(mask);
+        }
+    }
+
+    pub fn insert(&mut self, v: Datum) -> Result<()> {
```

Comment:
Use `&Datum` instead.

Reply:
Then how to convert from `&Datum` to `&[Datum]` in `encode_value` @BusyJay ?

Reply:
```rust
unsafe { mem::transmute::<_, &[Datum]>((&v, 1_usize)) }
```

🤣 never mind, a joke here.

Reply:
You can use `util::as_slice`.